Extend bat harness

srithon opened the pull request:
Instead of only printing out randomized bytes, I extended the harness to configure all parameters of the `PrettyPrinter`. The latest commit, "Implement Arbitrary directly on `bat` types", changes the implementation to directly generate an arbitrary `PrettyPrinter` rather than randomly calling methods on the builder. I did this to see if it would catch more exceptions, and it did. For some reason, this latest run was under a separate Mayhem project: https://mayhem.forallsecure.com/srithon/bat-mayhem, rather than the original https://mayhem.forallsecure.com/srithon/bat.

srithon committed:
Expand target to cover more PrettyPrinter options

srithon committed:
Remove commented-out fields from TestInput

These fields were mostly omitted because generating Arbitrary inputs of
their types would require adding dependencies to the main project, which
is undesireable.

srithon committed:
Add "arbitrary" dependency to fuzzing crate

srithon committed:
Remove "language" param.

This was causing too many to fail.

srithon committed:
Implement Arbitrary directly on `bat` types

xansec reviewed (changes requested):
It looks like this commit is based off an old upstream commit. Could you please sync your fork first? Please make the requested changes.

srithon commented:
> It looks like this commit is based off an old upstream commit. Could you please sync your fork first? Please make the requested changes.

Hmm, I looked at my fork and it seems that it's synced with the `mayhemheroes` repo. I also looked at the `sharkdp/bat` master branch, and `mayhemheroes` (and thus my fork) is synced with that as well; in other words, the HEAD of `sharkdp/bat` is an ancestor of HEAD on both my fork and on `mayhemheroes/bat`. Could you please elaborate on what needs to be synced?

Thanks

xansec commented:
Yes, you're right. I was surprised because I saw an an unusually high number of modified files, it appears now that this is just because you added the arbitrary decorator to a lot of the source. Is this necessary? In general, PRs should be additions, not modifications (as this makes it difficult to keep sync with upstream). Can you make your target and your PR without the additions? I have seen other rust submissions making use of the arbitrary decorator that don't modify upstream source. Same thing with cargo lock and toml - I would let upstream handle those files. Since it appears that you are just submitting an existing fuzz target and not necessarily adding new harnessing, you could easily get away with simply submitting a Mayhemfile, Dockerfile, and action.

srithon committed:
Revert "Implement Arbitrary directly on `bat` types"

This reverts commit bd5a47e8cbf60378a08099e1d5669a0262f8af47.

srithon commented:
> Yes, you're right. I was surprised because I saw an an unusually high number of modified files, it appears now that this is just because you added the arbitrary decorator to a lot of the source. Is this necessary? In general, PRs should be additions, not modifications (as this makes it difficult to keep sync with upstream). Can you make your target and your PR without the additions? I have seen other rust submissions making use of the arbitrary decorator that don't modify upstream source. Same thing with cargo lock and toml - I would let upstream handle those files. Since it appears that you are just submitting an existing fuzz target and not necessarily adding new harnessing, you could easily get away with simply submitting a Mayhemfile, Dockerfile, and action.

That's completely fair; I originally wrote my harness without modifying the source, and then in my latest commit, changed the implementation to call `Arbitrary` directly on the types within the crate, mainly to see if it would be faster.
I just reverted this commit since it's definitely intrusive and unmaintainable in the long run. Both implementations have the same end behavior, which is randomizing the `PrettyPrinter` configuration and seeing if it panics under any input.

xansec commented on fuzz/fuzz_targets/bat_stdin_input.rs:
```diff
@@ -5,15 +5,69 @@ use libfuzzer_sys::fuzz_target;
 /// This prints embedded bytes with a custom header and then reads from STDIN.
```
Sorry to be pedantic about this, but I would say if you are going to simply reuse the existing target, just submit a Mayhemfile, Dockerfile, and workflow file (all three seem to be missing from this PR). On the other hand, if you are going to update the target, why not just create a new one as a copy? A large percentage of submitted targets never end up fuzzing the most recent source because somewhere along the line a breaking commit was made, so I'm being picky here to make sure that happens as infrequently as possible.

xansec reviewed (changes requested):
Please make the requested changes.

xansec reviewed (approved):
LGTM!

xansec reviewed (changes requested):
accidental approval

srithon committed:
Revert changes to original harness, use new harness

This commit effectively reverts the following commits, and reapplies them to a
new file called `bat_stdin_input_random_options`:
- 25fd500
- 664ded8
- cc72450
- b1189d2

srithon committed:
Format new harness source file

srithon committed:
Add new harness to Dockerfile and Mayhemfile

srithon committed:
Add new harness as fuzz crate binary

srithon commented:
Sorry it's taken me so long to get back to this- I just finished my finals.
To sum up my changes:
1. I took your advice and restored the original harness
2. I put my modifications into a new fuzz target
3. I added the resultant binary to the Docker image under `/bat_stdin_input_random_options`
4. I added another `cmd` to the Mayhemfile, pointing to the new binary.

In hindsight, I believe my attempt to add another `cmd` to the Mayhemfile does not actually have the desired effect; I believed that this would add another fuzz target, but I don't think that's what it's doing. My current understanding is that Mayhemfile's describe single harnesses, and so I would have to make another one for my new harness. Is this understanding correct?

I also understand your concern about fuzz targets not compiling on HEAD. To fix this, then, should I completely move my new fuzz target outside of the original fuzz crate? My reasoning is that if at some point my target no longer compiled but the original target did, the overall `cargo fuzz build` may fail.

Lastly, should I make a separate Dockerfile for my new target, or could I have the same Dockerfile compile both fuzz targets and copy the binaries as needed?

Thanks

xansec commented:
> My current understanding is that Mayhemfile's describe single harnesses, and so I would have to make another one for my new harness. Is this understanding correct?

Yes. You can use multiple commands if you have multiple builds of the same target and want to take advantage of SE and instrumentation simultaneously.

> My reasoning is that if at some point my target no longer compiled but the original target did, the overall cargo fuzz build may fail.

Yes, that's fine. Sometime the fuzz targets stop building anyways due to upstream changes. 

> Lastly, should I make a separate Dockerfile for my new target, or could I have the same Dockerfile compile both fuzz targets and copy the binaries as needed?

I guess I didn't realize that the conflicts were due to someone else already integrating this target in the past. I don't think you need to worry so much about upstream changes in this case. The fuzz target hasn't been merged to upstream, so it will never be subject to downstream changes. I think you are ok using the same target/Dockerfile if it is an improvement.

srithon committed:
Revert changes 09d8b08..=3d3da02

This commit effectively returns HEAD to commit `8c61af2`.
At this point, these were the changes to the original fuzz target:

Instead of using only the default options for the `PrettyPrinter`,
randomize the options by building a `PrettyPrinter` instance
using the fuzz input, and feeding the created `PrettyPrinter` an
arbitrary string

srithon commented:
> > My reasoning is that if at some point my target no longer compiled but the original target did, the overall cargo fuzz build may fail.
> 
> Yes, that's fine. Sometime the fuzz targets stop building anyways due to upstream changes.
> 
> > Lastly, should I make a separate Dockerfile for my new target, or could I have the same Dockerfile compile both fuzz targets and copy the binaries as needed?
> 
> I guess I didn't realize that the conflicts were due to someone else already integrating this target in the past. I don't think you need to worry so much about upstream changes in this case. The fuzz target hasn't been merged to upstream, so it will never be subject to downstream changes. I think you are ok using the same target/Dockerfile if it is an improvement.

Sorry if we're still not on the same page, but my interpretation of these messages was that my initial implementation, minus the commit titled "Implement Arbitrary directly on `bat` types", was fine.
I got this from when you said "I think you are ok using the same target/Dockerfile if it is an improvement."

To clear up what this PR is about, here's a summary:

- In the original target (before my changes), the code was simply feeding arbitrary strings into a `PrettyPrinter`, and seeing if it caused a panic
- I wanted to also parameterize the `PrettyPrinter` using its custom options, so I created a `TestInput` struct containing optional values for each of the `PrettyPrinter` parameters, and fed them into the `PrettyPrinter` builder before making it print the arbitrary string
- This implementation fuzzes a superset of the original target's cases, since it not only passes in arbitrary strings to the `PrettyPrinter`, but it changes its configuration options as well

If anything is still unclear, or if I misinterpreted what you said, please let me know.

xansec reviewed (approved):
No, I think this assessment is correct. Sorry for the delay. LGTM!

srithon commented:
> No, I think this assessment is correct. Sorry for the delay. LGTM!

All good, thanks for the feedback!
